### PR TITLE
Add 358 release notes

### PR DESCRIPTION
**Pull request by martint**

### CODE REVIEW

File: docs/src/main/sphinx/release/release-358.md [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+* Reduce worker graceful shutdown time. ({issue}`8149`)
+* Fix query failure when one of join build side's columns is of non-orderable type. ({issue}`7723`)
+* Fix failure for queries containing repeated ordinals in `GROUP BY` clause.
+  Example: `SELECT x FROM t GROUP BY 1, 1`. ({issue}`8023`)
+* Fix failure for queries containing repeated expressions in `ORDER BY` clause of an aggregate function.
+  Example: `SELECT count(x ORDER BY y, y) FROM (VALUES ('a', 2)) t(x, y)`. ({issue}`8080`)  
+  
+## Security
+
+* Fix supports for multiple LDAP user bind patterns.  ({issue}`8022`)
```
Comment (martint) [inline docs/src/main/sphinx/release/release-358.md:17]: @dain, what exactly does this fix? Or is it more like "Add support for ..."?

File: docs/src/main/sphinx/release/release-358.md [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+## Hive connector
+
+* Add support for reading from Hive views which use `LATERAL VIEW EXPLODE`
+  or `LATERAL VIEW OUTER EXPLODE` over array of `STRUCTs`. ({issue}`8120`)
+  
+## Iceberg connector
+
+* Fix listing of materialized views in `information_schema.tables` and `system.jdbc.tables` ({issue}`8151`)
```
Comment (martint) [inline docs/src/main/sphinx/release/release-358.md:26]: @sopel39, what exactly does this fix? Is it a failure? Missing entries? Entries showing up when they shouldn't?

File: docs/src/main/sphinx/release/release-358.md (hunk 1 above)
Comment (hashhar) [reply docs/src/main/sphinx/release/release-358.md:17]: This was reverted - the correct fix is part of https://github.com/trinodb/trino/pull/8134

Comment (martint) [reply docs/src/main/sphinx/release/release-358.md:17]: Thanks, I'll remove the entry.

File: docs/src/main/sphinx/release/release-358.md (hunk 2 above)
Comment (martint) [reply docs/src/main/sphinx/release/release-358.md:27]: @raunaqmorarka, thoughts?

Comment (raunaqmorarka) [reply docs/src/main/sphinx/release/release-358.md:27]: Earlier there was query failure with message `Not an Iceberg table`

File: docs/src/main/sphinx/release/release-358.md [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+* Fix failure when listing materialized views in `information_schema.tables` and `system.jdbc.tables` ({issue}`8151`)
+
+## Memory connector
+
+* Improve performance of certain complex queries involving joins. ({issue}`8095`)
+
+## SPI
+
+* Remove deprecated ``ConnectorPageSourceProvider#createPageSource`` method overrides. ({issue}`8077`)
+* Add support for casting redirected table columns when source column types don't match. ({issue}`6066`)
```
Comment (sopel39) [inline docs/src/main/sphinx/release/release-358.md:36]: I think it should be
```
Add support for redirected table columns casting when source column types don't match
```

File: docs/src/main/sphinx/release/release-358.md [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+* Add `ConnectorMetadata#redirectTable` to allow the connectors to redirect table scans. ({issue}`7606`)
+* Add `ConnectorMetadata#streamTableColumns` for streaming columns in a redirection-aware manner. The
```
Comment (sopel39) [inline docs/src/main/sphinx/release/release-358.md:38]: `streaming columns` -> streaming column metadata`?

File: docs/src/main/sphinx/release/release-358.md [hunk 5]
```diff
@@ -0,0 +1,40 @@
+# Release 358 (XX May 2021)
+
+## General
+
+* Support any query in {doc}`/sql/show-stats`. ({issue}`8026`)
+* Improve performance of complex queries involving joins and `TABLESAMPLE`. ({issue}`8094`)
+* Improve performance of `ORDER BY ... LIMIT` queries on sorted data. ({issue}`6634`)
```
Comment (mosabua) [inline docs/src/main/sphinx/release/release-358.md:7]: Should we change to Top-N pushdown since that is the name we use in the docs now? https://trino.io/docs/current/optimizer/pushdown.html#top-n-pushdown

File: docs/src/main/sphinx/release/release-358.md [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+* Reduce worker graceful shutdown time. ({issue}`8149`)
```
Comment (mosabua) [inline docs/src/main/sphinx/release/release-358.md:8]: Reduce graceful shutdown time for workers.

or 

Reduce graceful shutdown time for worker nodes.

File: docs/src/main/sphinx/release/release-358.md [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+* Fix query failure when one of join build side's columns is of non-orderable type. ({issue}`7723`)
```
Comment (mosabua) [inline docs/src/main/sphinx/release/release-358.md:9]: Fix query failure when a build side column used in a join uses a non-orderable type.

File: docs/src/main/sphinx/release/release-358.md [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+* Fix failure for queries containing repeated ordinals in `GROUP BY` clause.
```
Comment (mosabua) [inline docs/src/main/sphinx/release/release-358.md:10]:
```
in the `GROUP BY` clause.
```

or 

```
in an  `GROUP BY` clause.
```

File: docs/src/main/sphinx/release/release-358.md [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+  Example: `SELECT x FROM t GROUP BY 1, 1`. ({issue}`8023`)
+* Fix failure for queries containing repeated expressions in `ORDER BY` clause of an aggregate function.
```
Comment (mosabua) [inline docs/src/main/sphinx/release/release-358.md:12]:
```
in the `ORDER BY` clause
```

or 

```
in an ORDER BY clause
```

File: docs/src/main/sphinx/release/release-358.md [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+  Example: `SELECT count(x ORDER BY y, y) FROM (VALUES ('a', 2)) t(x, y)`. ({issue}`8080`)  
+
+## JDBC Driver
```
Comment (mosabua) [inline docs/src/main/sphinx/release/release-358.md:15]: Add empty line below title

File: docs/src/main/sphinx/release/release-358.md [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+* Remove legacy driver classes `com.facebook.presto.jdbc.PrestoDriver`
```
Comment (mosabua) [inline docs/src/main/sphinx/release/release-358.md:16]: Is this a breaking change we need to call out more?

File: docs/src/main/sphinx/release/release-358.md [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+  and `io.prestosql.jdbc.PrestoDriver`. ({issue}`8042`)
+* Remove legacy JDBC URL prefix `jdbc:presto:`. ({issue}`8042`)
```
Comment (mosabua) [inline docs/src/main/sphinx/release/release-358.md:18]: Is this a breaking change we need to call out more?

File: docs/src/main/sphinx/release/release-358.md (hunk 11 above)
Comment (electrum) [inline docs/src/main/sphinx/release/release-358.md:16]: Nit: missing blank line

File: docs/src/main/sphinx/release/release-358.md [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+
+## Hive connector
+
+* Add support for reading from Hive views which use `LATERAL VIEW EXPLODE`
```
Comment (mosabua) [inline docs/src/main/sphinx/release/release-358.md:22]: views that use

File: docs/src/main/sphinx/release/release-358.md (hunk 3 above)
Comment (mosabua) [reply docs/src/main/sphinx/release/release-358.md:36]: table column casting ??

**Review by mosabua — CHANGES_REQUESTED** [review]
A couple of suggestions.

File: docs/src/main/sphinx/release/release-358.md [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+
+* Support any query in {doc}`/sql/show-stats`. ({issue}`8026`)
+* Improve performance of complex queries involving joins and `TABLESAMPLE`. ({issue}`8094`)
+* Improve performance of `ORDER BY ... LIMIT` queries on sorted data. ({issue}`6634`)
+* Reduce worker graceful shutdown time. ({issue}`8149`)
+* Fix query failure when one of join build side's columns is of non-orderable type. ({issue}`7723`)
+* Fix failure for queries containing repeated ordinals in `GROUP BY` clause.
+  Example: `SELECT x FROM t GROUP BY 1, 1`. ({issue}`8023`)
+* Fix failure for queries containing repeated expressions in `ORDER BY` clause of an aggregate function.
+  Example: `SELECT count(x ORDER BY y, y) FROM (VALUES ('a', 2)) t(x, y)`. ({issue}`8080`)
```
Comment (electrum) [inline docs/src/main/sphinx/release/release-358.md:13]: Maybe use `array_agg` since that function actually uses `ORDER BY`

File: docs/src/main/sphinx/release/release-358.md [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+* Reduce worker graceful shutdown time. ({issue}`8149`)
+* Fix query failure when one of join build side's columns is of non-orderable type. ({issue}`7723`)
+* Fix failure for queries containing repeated ordinals in `GROUP BY` clause.
+  Example: `SELECT x FROM t GROUP BY 1, 1`. ({issue}`8023`)
+* Fix failure for queries containing repeated expressions in `ORDER BY` clause of an aggregate function.
+  Example: `SELECT count(x ORDER BY y, y) FROM (VALUES ('a', 2)) t(x, y)`. ({issue}`8080`)  
+
+## JDBC Driver
+* Remove legacy driver classes `com.facebook.presto.jdbc.PrestoDriver`
+  and `io.prestosql.jdbc.PrestoDriver`. ({issue}`8042`)
```
Comment (electrum) [inline docs/src/main/sphinx/release/release-358.md:17]: Maybe we should flip these so the newer `io.prestosql` is first

File: docs/src/main/sphinx/release/release-358.md (hunk 12 above)
Comment (electrum) [inline docs/src/main/sphinx/release/release-358.md:18]: Let's put this first since JDBC URLs are more user facing than driver class names. (I know, these came from me originally...)

File: docs/src/main/sphinx/release/release-358.md [hunk 16]
```diff
@@ -0,0 +1,40 @@
...
+
+* Fix failure when listing materialized views in `information_schema.tables` and `system.jdbc.tables` ({issue}`8151`)
+
+## Memory connector
+
+* Improve performance of certain complex queries involving joins. ({issue}`8095`)
+
+## SPI
+
+* Remove deprecated ``ConnectorPageSourceProvider#createPageSource`` method overrides. ({issue}`8077`)
```
Comment (electrum) [inline docs/src/main/sphinx/release/release-358.md:35]: We should be consistent about this (and fix existing release notes), but I think it looks nicer to write these as `ConnectorPageSourceProvider.createPageSource()`. This better matches The Java Tutorials, etc. I don't feel strongly, but we should pick a style and document it in https://github.com/trinodb/trino/wiki/Release-Notes-Guidelines

File: docs/src/main/sphinx/release/release-358.md [hunk 17]
```diff
@@ -0,0 +1,40 @@
...
+* Add support for casting redirected table columns when source column types don't match. ({issue}`6066`)
+* Add `ConnectorMetadata#redirectTable` to allow the connectors to redirect table scans. ({issue}`7606`)
```
Comment (electrum) [inline docs/src/main/sphinx/release/release-358.md:37]: We already have `applyTableScanRedirect()` which is a very different thing, so we should avoid "table scan" here. Let's say 

> to allow connectors to redirect table reads and metadata listings.

File: docs/src/main/sphinx/release/release-358.md (hunk 3 above)
Comment (electrum) [inline docs/src/main/sphinx/release/release-358.md:36]: Let's say "redirected table scan" to differentiate from metadata redirection.

> Add support for casting the columns of a redirected table scan when source column types don't match.

File: docs/src/main/sphinx/release/release-358.md [hunk 18]
```diff
@@ -0,0 +1,40 @@
+# Release 358 (XX May 2021)
```
Comment (findepi) [inline docs/src/main/sphinx/release/release-358.md:1]: Jun

File: docs/src/main/sphinx/release/release-358.md (hunk 5 above)
Comment (findepi) [reply docs/src/main/sphinx/release/release-358.md:7]: I guess "sorted data" here refers to connector-sorted. Should this be under "Hive connector" section instead?

File: docs/src/main/sphinx/release/release-358.md (hunk 6 above)
Comment (findepi) [reply docs/src/main/sphinx/release/release-358.md:8]: Worth indicating whether this means reduced guarantees (eg smaller timeout value), or better logic.

File: docs/src/main/sphinx/release/release-358.md [hunk 19]
```diff
@@ -0,0 +1,40 @@
...
+
+## JDBC Driver
+* Remove legacy driver classes `com.facebook.presto.jdbc.PrestoDriver`
+  and `io.prestosql.jdbc.PrestoDriver`. ({issue}`8042`)
+* Remove legacy JDBC URL prefix `jdbc:presto:`. ({issue}`8042`)
+
+## Hive connector
+
+* Add support for reading from Hive views which use `LATERAL VIEW EXPLODE`
+  or `LATERAL VIEW OUTER EXPLODE` over array of `STRUCTs`. ({issue}`8120`)
```
Comment (findepi) [inline docs/src/main/sphinx/release/release-358.md:23]: move "s" outside of backticks and maybe reword to avoid plural.

File: docs/src/main/sphinx/release/release-358.md [hunk 20]
```diff
@@ -0,0 +1,40 @@
...
+  
+## Iceberg connector
+
+* Fix failure when listing materialized views in `information_schema.tables` and `system.jdbc.tables` ({issue}`8151`)
```
Comment (findepi) [inline docs/src/main/sphinx/release/release-358.md:27]: `system.jdbc.tables` is not really user-facing. `java.sql.DatabaseMetaData#getTables` is

**Review by findepi — APPROVED** [review]

File: docs/src/main/sphinx/release/release-358.md (hunk 5 above)
Comment (martint) [reply docs/src/main/sphinx/release/release-358.md:7]: No, we refer to it as ORDER BY ... LIMIT, as that's what users think about when they type their queries.

File: docs/src/main/sphinx/release/release-358.md (hunk 7 above)
Comment (martint) [reply docs/src/main/sphinx/release/release-358.md:9]: I'm actually thinking we should rephrase in a way that doesn't say "build side". That's an internal implementation concept, not something users can reason about.

File: docs/src/main/sphinx/release/release-358.md (hunk 5 above)
Comment (findepi) [reply docs/src/main/sphinx/release/release-358.md:7]: @mosabua topn pushdown is not involved here (no pushdown into the connector takes place in the case study's query)

@martint i unresolved for my other question

Comment (martint) [reply docs/src/main/sphinx/release/release-358.md:7]: > I guess "sorted data" here refers to connector-sorted. Should this be under "Hive connector" section instead?

It's an engine-level improvement. If a connector produces sorted data (and describes its properties to the engine), the engine will be able to take advantage of it. I guess we can add it to both.

File: docs/src/main/sphinx/release/release-358.md [hunk 21]
```diff
@@ -0,0 +1,42 @@
+# Release 358 (1 Jun 2021)
+
+## General
+
+* Support arbitrary queries in {doc}`/sql/show-stats`. ({issue}`8026`)
+* Improve performance of complex queries involving joins and `TABLESAMPLE`. ({issue}`8094`)
+* Improve performance of `ORDER BY ... LIMIT` queries on sorted data. ({issue}`6634`)
+* Reduce graceful shutdown time for worker nodes. ({issue}`8149`)
+* Fix query failure columns of non-orderable types are involved in a join. ({issue}`7723`)
```
Comment (losipiuk) [inline docs/src/main/sphinx/release/release-358.md:9]: Can we enumerate. Or at least give an example what a non-orderable type is?

File: docs/src/main/sphinx/release/release-358.md [hunk 22]
```diff
@@ -0,0 +1,42 @@
...
+
+## Hive connector
+
+* Add support for reading from Hive views that use `LATERAL VIEW EXPLODE`
+  or `LATERAL VIEW OUTER EXPLODE` over array of `STRUCT`. ({issue}`8120`)
+* Improve performance of `ORDER BY ... LIMIT` queries on sorted data. ({issue}`6634`)
+
+## Iceberg connector
+
+* Fix failure when listing materialized views in `information_schema.tables` and `java.sql.DatabaseMetaData.getTables()` ({issue}`8151`)
```
Comment (losipiuk) [inline docs/src/main/sphinx/release/release-358.md:29]:
```suggestion
* Fix failure when listing materialized views using query from `information_schema.tables` or JDBC API call to `java.sql.DatabaseMetaData.getTables()` ({issue}`8151`)
```

File: docs/src/main/sphinx/release/release-358.md [hunk 23]
```diff
@@ -0,0 +1,42 @@
...
+
+## Memory connector
+
+* Improve performance of certain complex queries involving joins. ({issue}`8095`)
+
+## SPI
+
+* Remove deprecated `ConnectorPageSourceProvider.createPageSource()` method overrides. ({issue}`8077`)
+* Add support for casting the columns of a redirected table scan when source column types don't match. ({issue}`6066`)
```
Comment (losipiuk) [inline docs/src/main/sphinx/release/release-358.md:38]:
```suggestion
* Add support for casting columns of a redirected table scan when source column types don't match. ({issue}`6066`)
```

File: docs/src/main/sphinx/release/release-358.md [hunk 24]
```diff
@@ -0,0 +1,42 @@
...
+* Add `ConnectorMetadata#redirectTable` to allow connectors to redirect table reads and metadata listings. ({issue}`7606`)
```
Comment (losipiuk) [inline docs/src/main/sphinx/release/release-358.md:39]:
```suggestion
* Add `ConnectorMetadata#redirectTable` to allow connectors redirect table reads and metadata listings. ({issue}`7606`)
```

File: docs/src/main/sphinx/release/release-358.md (hunk 23 above)
Comment (martint) [reply docs/src/main/sphinx/release/release-358.md:38]: I think "the" is fine in this sentence.

File: docs/src/main/sphinx/release/release-358.md (hunk 24 above)
Comment (martint) [reply docs/src/main/sphinx/release/release-358.md:39]: "to" is needed: "to allow connectors to <do something>"

File: docs/src/main/sphinx/release/release-358.md (hunk 5 above)
Comment (electrum) [reply docs/src/main/sphinx/release/release-358.md:7]: @mosabua we called it "Top-N pushdown" since we needed a succinct name to refer to that pushdown, and what it means is explained in that documentation section. We don't have a general term "Top-N" defined anywhere.

File: docs/src/main/sphinx/release/release-358.md [hunk 25]
```diff
@@ -0,0 +1,43 @@
...
```
Comment (electrum) [inline docs/src/main/sphinx/release/release-358.md:40]: `ConnectorMetadata.redirectTable()`

File: docs/src/main/sphinx/release/release-358.md [hunk 26]
```diff
@@ -0,0 +1,43 @@
...
+* Add `ConnectorMetadata#streamTableColumns` for streaming column metadata in a redirection-aware manner. The
```
Comment (electrum) [inline docs/src/main/sphinx/release/release-358.md:41]: `ConnectorMetadata.streamTableColumns()`

File: docs/src/main/sphinx/release/release-358.md [hunk 27]
```diff
@@ -0,0 +1,43 @@
...
+  alternate method for column listing `ConnectorMetadata#listTableColumns` is now deprecated. ({issue}`7606`)
```
Comment (electrum) [inline docs/src/main/sphinx/release/release-358.md:42]: `ConnectorMetadata.listTableColumns()`

**Review by electrum — APPROVED** [review]

File: docs/src/main/sphinx/release/release-358.md (hunk 5 above)
Comment (mosabua) [reply docs/src/main/sphinx/release/release-358.md:7]: Thanks for clarifying .. I clearly misunderstood.

Comment (martint) [reply docs/src/main/sphinx/release/release-358.md:7]: They can observe this with third party connectors, too, if they are running any. That’s why I think it also belongs in the general section. It’s not specific to the Hive connector.
